### PR TITLE
Implement simultaneous entanglement swapping

### PR DESCRIPTION
**Pull request by PorametPat**
Adding new rule, action, swappingConfig and etc. for simultaneous entanglement swapping situation,

### CODE REVIEW

File: quisp/modules/QRSA/ConnectionManager/ConnectionManager.cc [hunk 1]
```diff
@@ -348,7 +366,40 @@ void ConnectionManager::respondToRequest(ConnectionSetupRequest *req) {
       pkr->setApplication_type(0);
       send(pkr, "RouterPort$o");
     }
-  }
+    /*
```
Comment (zigen) [inline quisp/modules/QRSA/ConnectionManager/ConnectionManager.cc:369]: you should remove these comments because we are using git!

**Review by zigen — APPROVED** [review]

File: quisp/modules/QRSA/ConnectionManager/ConnectionManager.cc [hunk 2]
```diff
@@ -303,26 +306,42 @@ void ConnectionManager::respondToRequest(ConnectionSetupRequest *req) {
...
+        pkr->setDestAddr(path.at(i));
+        pkr->setSrcAddr(my_address);
+        pkr->setKind(2);
+        pkr->setRuleSet(rule);
+        pkr->setActual_srcAddr(path.at(0));
+        pkr->setActual_destAddr(path.at(path.size() - 1));
+        send(pkr, "RouterPort$o");
+      } else {
+        SwappingConfig config = generateSimultaneousSwappingConfig(path.at(i), path, qnics, num_resource);
+        RuleSet *rule = generateSimultaneousEntanglementSwappingRuleSet(path.at(i), config, path);
```
Comment (Chibikuri) [inline quisp/modules/QRSA/ConnectionManager/ConnectionManager.cc:334]: For the later implementation, this should be generalized. (Take "simultaneous" flag as the argument of generateEntanglementSwapppingConfig)

**Comment by Chibikuri** [comment]
Can you fix `Simultaneous swapping action` and resource update at RuleEngine with [action at e2e](https://github.com/sfc-aqua/quisp/blob/0dafb87ba481eba0a3a7b6553106e19ba1a78928/quisp/rules/Action.cc#L83) and [RuleEngine at e2e](https://github.com/sfc-aqua/quisp/blob/0dafb87ba481eba0a3a7b6553106e19ba1a78928/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L936)?

**Comment by PorametPat** [comment]
> Can you fix `Simultaneous swapping action` and resource update at RuleEngine with [action at e2e](https://github.com/sfc-aqua/quisp/blob/0dafb87ba481eba0a3a7b6553106e19ba1a78928/quisp/rules/Action.cc#L83) and [RuleEngine at e2e](https://github.com/sfc-aqua/quisp/blob/0dafb87ba481eba0a3a7b6553106e19ba1a78928/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L936)?

I'm not sure what the problems are?

**Comment by Chibikuri** [comment]
I think you followed my swapping action, but it doesn't have `freeConsumedResources` in the action. So I updated it like this https://github.com/sfc-aqua/quisp/blob/0dafb87ba481eba0a3a7b6553106e19ba1a78928/quisp/rules/Action.cc#L152
And also, when you add a entangled resource at allResource, you have to remove old records like this. https://github.com/sfc-aqua/quisp/blob/0dafb87ba481eba0a3a7b6553106e19ba1a78928/quisp/modules/QRSA/RuleEngine/RuleEngine.cc#L973

File: quisp/rules/Action.cc [hunk 3]
```diff
@@ -176,6 +176,108 @@ cPacket *SwappingAction::run(cModule *re) {
...
+    operation_type_left = 0;
+    operation_type_right = 3;
+  }
+  RuleEngine *rule_engine = check_and_cast<RuleEngine *>(re);
+  if (std::rand() / RAND_MAX < success_probability) {
+    right_partner_qubit->setEntangledPartnerInfo(left_partner_qubit);
+    left_partner_qubit->setEntangledPartnerInfo(right_partner_qubit);
+
+  } else {
+    left_partner_qubit->isBusy = false;
```
Comment (Chibikuri) [inline quisp/rules/Action.cc:236]: This part would be a bug, but we need e2e tomography to verify it.

**Review by Chibikuri — APPROVED** [review]
I have tested and so far it's working.
Thanks a lot! I'll check if it returns the proper result after I finish my e2e tomography.

**Pull request merged**